File: core/transformers/embed/normalize.js
function normalize(obj) {
	if (obj === undefined) {
		throw new Error("transformed value is undefined");
	}
	if (obj === null) {
		throw new Error("transformed value is null");
	}
	if (Array.isArray(obj)) {
		throw new Error("transformed value is array");
	}
	if (typeof obj !== "object") {
		throw new Error("transformed value is "+(typeof obj)+", not object");
	}
	function norm(obj, set) {
		if (set.has(obj)) {
			throw new Error("transformed value contains a circular reference");
		}
		set.add(obj);
		if (Array.isArray(obj)) {
			const len = obj.length;
			for (let i = 0; i < len; i++) {
				const v = obj[i];
				if (typeof v === "number") {
					if (!Number.isFinite(v)) {
						obj[i] = String(v);
					}
				} else if (v === undefined) {
					delete(obj[i]);
				} else if (typeof v === "object" && v !== null) {
					norm(v, set);
				} else if (typeof v === "bigint") {
					obj[i] = v.toString();
				}
			}
		} else {
			for (const k in obj) {
				if (obj.hasOwnProperty(k)) {
					const v = obj[k];
					if (typeof v === "number") {
						if (!Number.isFinite(v)) {
							obj[k] = String(v);
						}
					} else if (v === undefined) {
						delete(obj[k]);
					} else if (typeof v === "object" && v !== null) {
						norm(v, set);
					} else if (typeof v === "bigint") {
						obj[k] = v.toString();
					}
				}
			}
		}
		set.delete(obj);
	}
	norm(obj, new WeakSet());
}

File: core/internal/transformers/embed/normalize.py
class _Norm:
    @staticmethod
    def _infinite_to_string(n):
        import math

        if math.isnan(n):
            return "NaN"
        if n > 0:
            return "Infinite"
        return "-Infinite"

    @staticmethod
    def normalize(obj: dict) -> None:
        import math

        if type(obj) is not dict:
            raise ValueError(f"transformed value is '{type(obj).__name__}', not 'dict'")

        def norm(obj: dict, s: set) -> None:
            if id(obj) in s:
                raise ValueError("transformed value contains a circular reference")
            s.add(id(obj))
            if type(obj) is dict:
                for k, v in obj.items():
                    if type(v) in [dict, list]:
                        norm(v, s)
                    elif type(v) is float and not math.isfinite(v):
                        obj[k] = _Norm._infinite_to_string(v)
            elif type(obj) is list:
                for i, v in enumerate(obj):
                    if type(v) in [dict, list]:
                        norm(v, s)
                    elif type(v) is float and not math.isfinite(v):
                        obj[i] = _Norm._infinite_to_string(v)
            s.remove(id(obj))

        norm(obj, set())
